omod/src/main/java/org/openmrs/module/adminui/page/controller/myaccount/ChangeSecretQuestionPageController.java

```diff
@@ -9,98 +9,72 @@
-
-        public void setPassword(String password) {
-            this.password = password;
-        }
-
-        public String getQuestion() {
-            return question;
-        }
-
-        public void setQuestion(String question) {
-            this.question = question;
-        }
-
-        public String getAnswer() {
-            return answer;
-        }
-
-        public void setAnswer(String answer) {
-            this.answer = answer;
-        }
-
-        public String getConfirmAnswer() {
-            return confirmAnswer;
-        }
-
-        public void setConfirmAnswer(String confirmAnswer) {
-            this.confirmAnswer = confirmAnswer;
-        }
-
-        @Override
-        public String toString() {
-            return "Question=" + question + ", Answer=" + answer;
-        }
-    }
+	
+	public void get(PageModel model) {
+		String secretQuestion = fetchExistingSecretQuestionOrEmpty();
+		model.put("secretQuestion", secretQuestion);
+	}
+	
+	public String post(PageModel model, @RequestParam("password") String password,
+	                   @RequestParam("question") String question, @RequestParam("answer") String answer,
+	                   @RequestParam("confirmAnswer") String confirmAnswer,
+	                   @SpringBean("userService") UserService userService, HttpServletRequest request) {
+		if (!answer.equals(confirmAnswer)) {
+			model.put("secretQuestion", fetchExistingSecretQuestionOrEmpty());
+			request.getSession().setAttribute(UiCommonsConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
+			    Context.getMessageSourceService().getMessage("adminui.account.answerAndConfirmAnswer.doesNotMatch"));
+			
+			return "myaccount/changeSecretQuestion";
+		} else {
+			return changeSecretQuestion(password, question, answer, userService, request, model);
+		}
+	}
+	
+	private String changeSecretQuestion(String password, String question, String answer, UserService userService,
+	                                    HttpServletRequest request, PageModel model) {
+		try {
+			userService.changeQuestionAnswer(password, question, answer);
+			InfoErrorMessageUtil.flashInfoMessage(request.getSession(),
+			    Context.getMessageSourceService().getMessage("adminui.account.secretQuestion.success"));
+		}
+		catch (Exception ex) {
+			request.getSession().setAttribute(
+			    UiCommonsConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
+			    Context.getMessageSourceService().getMessage("registrationapp.save.fail") + "<br />"
+			            + ex.getLocalizedMessage());
+			model.put("secretQuestion", question);
+			
+			return "myaccount/changeSecretQuestion";
+		}
+		return "myaccount/myAccount";
+	}
+	
+	public String fetchExistingSecretQuestionOrEmpty() {
+		LoginCredential credential = getComponent(UserDAO.class).getLoginCredential(Context.getAuthenticatedUser());
+		String secretQuestion = credential.getSecretQuestion();
+		
+		return secretQuestion == null ? "" : secretQuestion;
+	}
+	
+	private <T> T getComponent(Class<T> clazz) {
```

Comment:
could be better mentioning it here so that i can replace this. My worry was whether is not a security compromise to put "LoginCredential credential = getComponent(UserDAO.class).getLoginCredential(Context.getAuthenticatedUser());" i.e. call into UserDAO class from this or any controller!